src/messageCache.js

```diff
@@ -0,0 +1,115 @@
+const RPC = require('./message').rpc.RPC
+const utils = require('./utils')
+/**
+ * This file implements the Message Cache API provided in https://github.com/libp2p/go-libp2p-pubsub/blob/master/mcache.go#L15 used in gossip sub to store messages that were sent for the last few heartbeat ticks.
+ */
+
+
+class CacheEntry {
+
+    /**
+     * @param {String}
+     * @param {String[]}
+     * @constructor
+     */
+    constructor (msgID, topics) {
+        this.msgID = msgID
+	this.topics = topics
+    }
+
+}
+
+class MessageCache {
+
+    /**
+     * @param {Number} gossip
+     * @param {Number} history
+     *
+     * @constructor
+     */
+    constructor (gossip, history) {
+	/**
+	 * @type {Map<string, RPC.Message>}
+	 */
+        this.msgs = new Map()
+
+	/**
+	 * @type {Array<Array{CacheEntry}>}
+	 */
+	this.history = []
+	for (let i=0; i < history; i++){
+	     this.history.push([])
+	}
+	
+	/**
+	 * @type {Number}
+	 */
+	this.gossip = gossip
+    }
+
+    /**
+     * Adds a message to the current window and the cache
+     *
+     * @param {pb.rpc.RPC.Message Object}
+     *
+     */
+    put (msg) {
+	var msgID = utils.msgId(msg.from, msg.seqno)
```

Comment:
I inconsistently used `var` in this file only. I tend to slip into poor habits when learning something difficult. I will be more consistent in the future.

As for your second point, I'm not quite sure I understand. I copied over the `utils.js` from `js-libp2p-floodsub`. The definition of `utils.msgId` is such that you need to pass in both `msg.from` and `msg.seqno`. This can very easily be amended later on.

Reply:
No problem, just for making sure everything gets great!

We can probably move the utils to `js-libp2p-pubsub` and refactor it to my suggestion. What do you think? Then, both floodsub and gossipsub require the the common utils from there